idelib/util.py

```diff
@@ -35,80 +37,154 @@ def verify(data, schema=None):
+                        blockEnd = subEl.value
+
+                blockEnd = blockEnd or blockStart
+                if chId is None:
+                    logger.warning("Extractor: {} missing <ChannelIDRef> subelement, skipping.".format(el))
+                    continue
+                if blockStart is None:
+                    logger.warning("Extractor: {} missing <StartTimeCodeAbs> subelement, skipping.".format(el))
+                    continue
+
+                if channels and chId not in channels:
+                    continue
+                if finished.setdefault(chId, False):
+                    continue
+
+                # TODO: Modulus correction, if still needed.
+                scalar = timeScalars.get(chId, 1)
+                blockStart *= scalar
+                blockEnd *= scalar
+
+                writeCurrent = True
+                writePrev = False  # write previous block, if current one starts late
+
+                if blockEnd < startTime:
+                    # Entirely before extraction interval. Write nothing.
+                    writeCurrent = False
+                elif blockStart <= startTime:
+                    # Block overlaps start of interval. Write block.
+                    # Mark channel finished if block also includes end of interval.
+                    writePrev = False
+                    finished[chId] = blockEnd >= endTime
+                elif blockEnd <= endTime:
+                    # Block within interval. Write block (w/ prev. if needed).
+                    writePrev = channelsWritten[chId] == 0
+                else:
+                    # Block overlaps end of interval. Write block (w/ prev. if needed).
+                    # Mark channel finished.
+                    finished[chId] = True
+                    writePrev = channelsWritten[chId] == 0
+
+                if writePrev:
+                    # Write the previous block, which is before the extraction interval.
+                    # This is to ensure that initial data is not left out.
+                    prev = lastBlocks.get(chId, None)
+                    if prev:
+                        data = prev[0].getRaw()
+                        fs.write(data)
+                        copiedBytes += len(data)
+                        channelsWritten[chId] += 1
+
+                lastBlocks[chId] = (el, blockStart, blockEnd)
+
+                if writeCurrent:
+                    channelsWritten[chId] += 1
+                else:
+                    # Skip to next element without writing anything.
+                    continue
+
+            data = el.getRaw()
+            fs.write(data)
+            copiedBytes += len(data)
+
+    except (StopIteration, KeyboardInterrupt):
+        pass
+
+    finally:
+        if isinstance(out, (str, Path)):
+            fs.close()
+
+    if updater:
+        updater(done=True)
+
+    return copiedBytes, sum(channelsWritten.values())
```

Comment:
Would it be preferable to ask the file stream how many bytes have been written rather than keep track?

```suggestion
    return fs.tell(), sum(channelsWritten.values())
```

Reply:
Since it's already written, I don't see much of an advantage in changing it.... also, what if the stream it was passed for output didn't start at `fs.tell() == 0`, or it is a file-like object without `tell()`?

Reply:
Hadn't thought that it might be a stream-like that had already been written to, that's a good point.